iceberg/iceberg-handler/src/main/java/org/apache/iceberg/mr/hive/compaction/IcebergMajorQueryCompactor.java

```diff
@@ -44,22 +52,68 @@ public boolean run(CompactorContext context) throws IOException, HiveException,
     Map<String, String> tblProperties = context.getTable().getParameters();
     LOG.debug("Initiating compaction for the {} table", compactTableName);
 
-    String compactionQuery = String.format("insert overwrite table %s select * from %<s",
-        compactTableName);
+    String partSpec = context.getCompactionInfo().partName;
+    String compactionQuery;
+
+    if (partSpec == null) {
+      HiveConf.setVar(context.getConf(), ConfVars.REWRITE_POLICY, RewritePolicy.ALL_PARTITIONS.name());
+      compactionQuery = String.format("insert overwrite table %s select * from %<s", compactTableName);
+    } else {
+      HiveConf.setVar(context.getConf(), ConfVars.REWRITE_POLICY, RewritePolicy.DEFAULT.name());
+      Table table = IcebergTableUtil.getTable(context.getConf(), context.getTable());
+      PartitionData partitionData = DataFiles.data(table.spec(), partSpec);
+      context.getConf().set(Context.compactPartitionSpec, partSpec);
+      compactionQuery = String.format("insert overwrite table %1$s partition(%2$s) select * from %1$s where %3$s",
+          compactTableName, partDataToSQL(partitionData, partSpec, ","),
+          partDataToSQL(partitionData, partSpec, " and "));
+    }
 
     SessionState sessionState = setupQueryCompactionSession(context.getConf(),
         context.getCompactionInfo(), tblProperties);
-    HiveConf.setVar(context.getConf(), ConfVars.REWRITE_POLICY, RewritePolicy.ALL_PARTITIONS.name());
+
     try {
       DriverUtils.runOnDriver(context.getConf(), sessionState, compactionQuery);
       LOG.info("Completed compaction for table {}", compactTableName);
+      return true;
     } catch (HiveException e) {
-      LOG.error("Error doing query based {} compaction", RewritePolicy.ALL_PARTITIONS.name(), e);
-      throw new RuntimeException(e);
+      LOG.error("Failed compacting table {}", compactTableName, e);
+      throw e;
     } finally {
       sessionState.setCompaction(false);
     }
+  }
+
+  private String partDataToSQL(PartitionData partitionData, String partSpec, String delimiter)
```

Comment:
you should use logic from `truncatePartitionBySpec` as you can't override partition with non identity spec

if (!TransformSpec.TransformType.IDENTITY.equals(transformType)) {
              throw new MetaException(
                  String.format("Partition transforms are not supported in compact operation: %s", entry.getKey()));
            }

Reply:
Done